### ISSUE TITLE
Behavior of Netty on Hanging Remotely-closed Proxy Connection

### ISSUE THREAD

**Issue by eaglerainbow** [netty/netty#8685]
In https://github.com/promregator/promregator/issues/83 I am currently struggling with outbound HTTP connections which use HttpProxyHandler. Scenario is as following:

* Promregator is a client application which uses https://github.com/cloudfoundry/cf-java-client
* cf-java-client in turn is using netty for making outbound TCP HTTP connections.
* In the scenario, an HTTP proxy must be used to reach the server endpoint on the other end. No authentication, no SOCKS...
* Promregator is creating HTTP requests periodically - usually every 15-60 seconds.
* Obviously proxy connections get pooled/reused by netty.

Initially the setup works fine and hundreds of requests can be sent successfully. After some time (usually around 1h), the entire application start to hang, though. The log statements indicate that no further connections can be made to the server.
By chance I could observe the network wire in such a situation and observed that the TCP connection between the application and the proxy server was closed by the proxy. The application considered the socket to still be open. The assumption is that the proxy simply dropped the connection in an unfriendly way (e.g. the proxy has a "maximal time to live" timeout for open connections).

I tried to read through the docs and many of the issues related to the search term "proxy" here, but I could not find any documentation describing how netty is expected to handle this situation (but perhaps I am simply not deep enough in the framework - NB I am from high up in OSI level 7 😁 )

What is the expected behavior based on your point of view for this scenario?
Should this situation be handled by netty or what would an application have to do to properly handle the  situation?
Any ideas which loggers I could activate to assist in analysis?

### Actual Behavior
The closed proxy connection apparently is kept open and obviously blocks resources. No further connections are processed.

### Expected Behavior
The socket of the closed proxy connection should be closed properly and - if needed - a new connection to the proxy shall established for the "broken" one.
(it is unclear to me, whether this is a task for netty or for the consumer of netty - in this case here: cf-java-client)

### Netty version
4.1.31

### JVM version (e.g. `java -version`)
```
openjdk version "1.8.0_181"
OpenJDK Runtime Environment (build 1.8.0_181-8u181-b13-2~deb9u1-b13)
OpenJDK 64-Bit Server VM (build 25.181-b13, mixed mode)
```
### OS version (e.g. `uname -a`)
``` 
Linux 8aeb66cfbf47 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 GNU/linux
```

**Comment by SharpMan** [comment]
I doubt this is a Netty issue. As you mentioned in the parent issue

> Netty has pooled the connections and tries to reuse all the connections over and over again (instead of trying to reopen the connections).

This is the main reason I can think of. Your clients running that has timed out may have issues when being called to be closed/timed-out (Probably an error is thrown  and is not being caught, an inaccurate buffer is miss-handled, or the connection might be marked closed but it's not) . 
I suggest you debug the client class and check what happens line by line when a client is being closed.

**Comment by eaglerainbow** [comment]
Thanks for the response!
In the meantime I could determine that netty isn't pooling connections actively (using a keep-alive technique). Each time a connection is made freshly (see also the "PCAP trace section" below). 

Debugging is quite tricky: I already tried to do it - and my limited knowledge about netty also isn't helping here either. Moreover, all the coding is in reactor style, so you have inversion of control all over the place. Moreover, the number of threads are quite high, so you get lost in all the breakpoints which get hit...

For the sake of analysis, I have implemented a logic, which tries to throw away the entire cf-java-client (which includes the reference to the netty `HttpClient` instance) in case that a test probe does not respond within a timeout of 2500ms (which is ages compared to the usual response time of a probing request). This "watchdogging" is performed every two minutes. The logs of the watchdog indicates that during the time of one outage (which took longer than 6 minutes), it was necessary to reset the connection three times in a row (each with 2 minutes delay) before the processing was fine again. I ran a PCAP trace in parallel which showed:
* In the "outage phase", netty still sends HTTPS requests through the proxy (e.g. our watchdog request), the connection is established successfully, application data is transferred back and forth (I can't see details though, because the data is TLS1.2 encrypted) and the connection is closed by the local station properly (SYN packet). The application data coming from the remote also was ACKed by the local station.
* The watchdog request in the middle took 180ms end-2-end (including setup of the proxy channel). 
* Even 2300ms later the data was not received by the application yet (and that's why the latter timed out). 
* During that time frame, no other netty connection was sent (most likely also not requested by the application).

The machine also wasn't overloaded: During that timeframe, short-term loadavg was in the interval `[0.07, 0.4]` (having 2 cores).
If it was just a pooling issue/buffer issue/resource leak, using a completely new `HttpClient` should have given us a completely new set of resources. So, I would have expected to see one watchdog iteration to fail - but it should have been fixed already with the next one after two minutes (where it wasn't). 
A network/server issue may also be ruled out, as then I would not have seen the corresponding ACK packets.
Also the "SYN" packet generated by the local station is speaking against the theory that closing the connection could be the source of error (I also wouldn't know how I could do a mistake in closing the client/connections when using netty-reactor). 

What puzzles me even more is the following observation: I have two instances of the application running - the first is the one which is causing the troubles here. The other is configured only with the following two deviations:
* It does not use a proxy configuration for netty.
* It connects to a different target (server), but the type of requests, the type of data returned, the timing, etc. is the same. 

(Note that it is technically not possible to make the first instance connect to its target without a proxy 😢 ). With the second instance, I never experienced any kind of issue like this - only the first one is affected.
Oh, and before anyone asks: Yes, it is reproducible over multiple restarts. Each instance is running in a docker container, which I already have recreated a dozen times. For me, only the time when the outage takes place is unpredictable.

I will keep trying to analyze the issue, but it's very complex as I don't have the necessary (detailed framework) knowledge for properly doing so. Thus, hints on debugging aids etc. are welcome.

**Comment by eaglerainbow** [comment]
Over the course of the day, I did some more debugging. I summarized my findings in
* https://github.com/promregator/promregator/issues/83#issuecomment-449968919 and
* https://github.com/promregator/promregator/issues/83#issuecomment-450022986

@SharpMan I would be greatful if you could read my analysis there and give me some hints, perhaps what else I could do to enable some more logging - as said: debugging the problem is very cumbersome.

**Comment by eaglerainbow** [comment]
I think I found the root cause in a cache, which is on application level. 

However, I also got indications that under high-load situation there is still a race condition (in 4.1.31), which could lead to the debug-leveled error message

``` 
IllegalArgumentException: Channel [id: xxxxxxx] was not acquired from this ChannelPool
``` 

which is raised by https://github.com/netty/netty/blob/9e5073960191d43dfa07277545f272da0c1f385a/transport/src/main/java/io/netty/channel/pool/SimpleChannelPool.java#L300

However, I could not find a reproducible case, which was adequate for being posted here.

**Comment by SharpMan** [comment]
@eaglerainbow 
Something closes the channel twice before it finishes releasing the first call. 
What was the fix done on the previous issue ?

**Comment by eaglerainbow** [comment]
@SharpMan 

> Something closes the channel twice before it finishes releasing the first call.

Well, there is little asking to close it (at least how far I am able to overlook all this complex stuff). I have more the impression that the system "thinks" that (due an error?) the connection has to be closed. But this is mainly just guess-work from my side now.

> What was the fix done on the previous issue ?

That's a little more complex, but I'll try to summarize it as much as possible:
* Promregator (the client where netty is running on) sends several requests to a server using the HTTPS request at very regular intervals. There is no eventing/WebSocket approach in place, so it's mainly scheduled polling. The connections all go against the same host (well, which is load-balanced to survive all the load...).
* The different types of information requested have a lifetime usually somewhere between 60s and 1hr. 
* The requests sent are partly quite costly (for the server). That is why Promregator tries to cache the requests during their lifetime as much as possible - the caching logic involved is quite tricky.
* Due to a bug in this caching logic, it ended up that for each request to the cache still a request to the HTTPS server was triggered - even though the answer was (in most cases) provided from the cache. 
* This led to hundreds of requests within a short timeframe (usually within 15 seconds). 
* My observations had been that "some very low number (<1%) of these requests" somehow went a-miss - and never returned (the cache has a timeout check in place, which logs the case that a re-fetch from the server takes too long). The discussions above a mirroring parts of my analysis there.

From the network traces I took, it could be seen that the problem must be located "somewhere on the client side". At the same time, however, I could prove that the requests had been sent from my application coding properly (I saw corresponding records in my logs) - and that had worked well a hundred times before!

Having detected the caching issue, it was immediately clear that this bug has to be fixed first. The fix then caused that the large number of requests decreased drastically, as the cache finally started doing its job properly again. With that lower number of requests, I could not reproduce the original netty/network problem anymore - even after a runtime of several days (which was never achievable before). That is why I came to the conclusion that there is "something rotten", if a massive number of connection attempts are being made in a short timeframe (high-load situation) - some of these apparently don't get through properly (to whatever reason). My experience tells me that this is most likely some awkward race condition, which nobody could ever think of until you see the proof in your logs yourself.

As interest here seemed to be not very high for finding the root cause and my problem was practically solved, I closed the issue here -- even though my impression is that "there is something out there".
